Eui image focus states

miukimiu opened the pull request:
### Summary

Closes #1983 

Added hover and focus states when `allowFullScreen` is true in EuiImage. 

Non-clickable images were being wrapped by a button so I removed the button for images that aren't clickable. 

### Checklist

- [x] Checked in **dark mode**
- [ ] Checked in **mobile**
- [x] Checked in **IE11** and **Firefox**
- [ ] Props have proper **autodocs**
- [ ] Added **documentation** examples
- [x] Added or updated **jest tests**
- [ ] Checked for **breaking changes** and labeled appropriately
- [ ] Checked for **accessibility** including keyboard-only and screenreader modes
- [x] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately

cchaos commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 ## [`master`](https://github.com/elastic/eui/tree/master)
 
 - Added href prop to `EuiTab` and converted to TypeScript ([#2275](https://github.com/elastic/eui/pull/2275))
+- Added hover and focus states when allowFullScreen is true in `EuiImage`([#2287](https://github.com/elastic/eui/pull/2287))
```
Because `allowFullScreen` is property we like to put that in a code block as well
```suggestion
- Added hover and focus states when `allowFullScreen` is true in `EuiImage`([#2287](https://github.com/elastic/eui/pull/2287))
```

cchaos commented on src/components/image/_image.scss:
```diff
@@ -17,9 +17,26 @@
     }
   }
 
-  &.euiImage--allowFullScreen:hover {
```
So I noticed that by removing the `:hover` from this outer selector, the `. euiImage__icon` is not always visible instead of just on hover. Was that on purpose?

**Before on the left**
<img src="https://d.pr/free/i/bzlGMF+" />

miukimiu replied:
It was already like that so I tried to keep as it was. But I can change. Do you think is better to leave the icon always visible so the users will know that it's clickable?

cchaos replied:
The problem with leaving it visible is that it obscures the image. So it should only show on hover and focus.

cchaos commented on src/components/image/image.js:
```diff
@@ -85,59 +87,56 @@ export class EuiImage extends Component {
+              ref={node => {
+                this.figure = node;
+              }}
+              className="euiImageFullScreen">
```
For BEM naming conventions, since `FullScreen` is actually a state, you should use a single dash like:
```suggestion
              className="euiImage-isFullScreen">
```

miukimiu replied:
Thanks for the suggestion. This class was part of the old code and I was not sure if I should change it. I updated it with the BEM naming convention and the corresponding SASS file.

cchaos commented on src/components/image/image.js:
```diff
@@ -85,59 +87,56 @@ export class EuiImage extends Component {
+                this.figure = node;
+              }}
+              className="euiImageFullScreen">
+              <img src={url} className="euiImageFullScreen__img" alt={alt} />
```
Same for this class
```suggestion
              <img src={url} className="euiImage-isFullScreen__img" alt={alt} />
```

myasonik commented:
How'd you feel about tackling some of the other a11y issues with images as well? 🙏 

**TL;DR** my proposed final markup (including the optional `figcaption` and full screen button) is:
```
<div>
	<figure role="figure" aria-label={figcaption}>
		<img src={src} alt={alt} />
		<figcaption>{figcaption}</figcaption>
	</figure>
	
	<button aria-label=`Show ${alt} image full screen`>
		{fullScreenIcon}
	</button>
</div>
```

#### So, why these changes?
1. `figure` elements aren't allowed inside of `button` elements which makes it tricky to get to the image for screen readers (you can try it out and get more info about it from [w3c's nu validator](https://validator.w3.org/nu/#textarea)
2. Figure needs the `role` fallback for IE11 and the `aria-label` for some screen reader/browser combinations to be accessible ([Scott O' Hara's if you got more details](https://www.scottohara.me/blog/2019/01/21/how-do-you-figure.html))
3. The `button` will now always tell screen readers what it does

#### Other notes
1. You can move the focus state to the containing `div` using `:focus-within` for supporting browsers (excludes IE11 and Edge, though Edge@next supports it, so you'll still need a workable fallback)
2. I 100% agree with you that the drop-shadow is way to subtle. I had to go double check that I was on the right branch because I thought there was no focus state and then I had to go check in the dev tools to see that it was being applied.
3. If there's no `button` you can get rid of the surrounding `div` as well though maybe that's more effort than it's worth. If there's no `figcaption` probably removing the `figure` is the right call all together though it is hard to discern implementing developer intent. If there's no `figcaption` or `button` you could remove everything except the `img` which is now a pretty substantial simplification of the DOM so maybe all of this _is_ worth it after all.
4. Not included in my a11y improvements suggestions is any mention of that our focus trap isn't perfect for full screen takeovers like this but I think that's a big enough problem left for later (#502).

cchaos commented:
Thanks @myasonik ! However, I think all of that is out of scope of this PR. This PR is mainly to just address the missing `:hover, :focus` states. Would you mind, however, creating an issue with your findings and proposed solutions?

miukimiu commented:
@cchaos when I've seen your comment I was almost finishing the a11y suggestions. 

@myasonik and @cchaos here's what I implemented so far:

1. I removed the figure elements from inside the button.
2. I moved the focus state to the containing div using `:focus-within`. Now we need to support (IE11 and Edge). I was wondering how I can tackle this. Should I create a new issue? 
3. I simplified the structure so when there's no `figcaption` there's no `figure` surrounding the `image`. But there's a `div` to handle some required classes.

@cchaos can you review the new focus and hover states?

cchaos commented on src/components/image/_image.scss:
```diff
@@ -71,20 +104,47 @@
   }
+
+  &:focus-within .euiImage__img {
+    border: 2px solid $euiFocusRingColor;
```
Because you're only adding a border size to the focus state, the focus state makes the image size and therefore the content around it jump.

<img src="https://d.pr/free/i/V4Bq6u+" width="50%" />

To solve this kind of issue, there are two options:

1. Add a transparent border with the same thickness to all states and then just change the color on focus.
2. Use the `outline` property instead.

I'd opt for method two or else when consumers want their image to reach all the way to it's container's bounds, it would never reach because of the border.

miukimiu replied:
I used the outline. Thanks for the suggestion! 👍

cchaos commented on src/components/image/_image.scss:
```diff
@@ -11,21 +11,58 @@
 
     .euiImage__icon {
+      // transition the shadow
+      transition: all $euiAnimSpeedFast $euiAnimSlightResistance;
```
There is no shadow on the icon, I think the comment is supposed to be opacity in which the property here should change to that.
```suggestion
      transition: opacity $euiAnimSpeedFast $euiAnimSlightResistance;
```

miukimiu replied:
I found out the issue was because the icon is actually an SVG and doesn't support `opacity`. So I changed to `fill-opacity` and it's working.

cchaos commented on src/components/image/_image.scss:
```diff
@@ -50,14 +87,10 @@
 .euiImage__caption {
   @include euiFontSizeS;
   text-align: center;
+  transition: all $euiAnimSpeedFast $euiAnimSlightResistance;
```
There isn't anything that changes on the caption that can be transitioned. I'd remove this.

miukimiu commented:
@cchaos and @myasonik I reverted to structure to what it was before the a11y changes. I also created a new branch [here](https://github.com/miukimiu/eui/tree/euiImage-focus-states-a11y), so we can have the accessibility discussion there. I'm going to create an issue for that.

cchaos commented on src/components/image/_image.scss:
```diff
@@ -17,14 +17,40 @@
+    &:hover .euiImage__icon {
       visibility: visible;
-      opacity: 1;
+      fill-opacity: 1;
```
My only worry is that for some reason the compatibility chart for this property is quite blank. So just make absolute sure you check in IE11, Edge, Safari, Firefox (and Chrome).

<img width="1127" alt="Screen Shot 2019-09-13 at 14 44 43 PM" src="https://user-images.githubusercontent.com/549577/64886861-2ee6d000-d635-11e9-8129-e8a98a3d5c14.png">

miukimiu replied:
I tested on different browsers IE11, Edge, Safari and Firefox and it seems t be working fine.

cchaos commented on src/components/image/_image.scss:
```diff
@@ -17,14 +17,40 @@
-      cursor: pointer;
+  &.euiImage--allowFullScreen {
+    &:focus .euiImage__img {
+      outline: 2px solid $euiFocusRingColor;
```
Much better 🎉

cchaos commented on src/components/image/image.js:
```diff
@@ -85,59 +87,62 @@ export class EuiImage extends Component {
     }
 
-    let optionalIcon;
+    const allowFullScreeIcon = (
```
Typo
```suggestion
    const allowFullScreenIcon = (
```

cchaos commented on src/components/image/image.js:
```diff
@@ -85,59 +87,62 @@ export class EuiImage extends Component {
+          <figure {...rest}>
+            <img src={url} className="euiImage__img" alt={alt} />
+            {optionalCaption}
+            {allowFullScreeIcon}
```
```suggestion
            {allowFullScreenIcon}
```

thompsongl commented:
@miukimiu #2328 merged, so I'll get the conflicts sorted for you

thompsongl commented:
jenkins test this
